pytest_inmanta/plugin.py

```diff
@@ -384,59 +400,64 @@ class Project:
     environment variable. Repositories are separated with spaces.
     """
 
-    def __init__(self, project_dir, load_plugins: bool = True):
+    def __init__(self, project_dir: str, load_plugins: bool = True) -> None:
         self._test_project_dir = project_dir
-        self._stdout = None
-        self._stderr = None
-        self.types = None
-        self.version = None
-        self.resources = {}
-        self._root_scope = {}
-        self._exporter = None
-        self._blobs = {}
-        self._facts = defaultdict(dict)
+        self._stdout: typing.Optional[str] = None
+        self._stderr: typing.Optional[str] = None
+        self.types: typing.Optional[typing.Dict[str, inmanta.ast.Type]] = None
+        self.version: typing.Optional[int] = None
+        self.resources: ResourceDict = {}
+        self._root_scope: typing.Optional[inmanta.ast.Namespace] = None
+        self._exporter: typing.Optional[Exporter] = None
+        self._blobs: typing.Dict[str, bytes] = {}
+        self._facts: typing.Dict[
+            ResourceIdStr, typing.Dict[str, typing.Any]
+        ] = defaultdict(dict)
         self._load()
-        self._plugins: Optional[Dict[str, object]] = (
+        self._plugins: typing.Optional[typing.Dict[str, FunctionType]] = (
             self._load_plugins() if load_plugins else None
         )
-        self._capsys = None
-        self.ctx = None
-        self._handlers = set()
+        self._capsys: typing.Optional[CaptureFixture] = None
+        self.ctx: typing.Optional[HandlerContext] = None
+        self._handlers: typing.Set[ResourceHandler] = set()
         config.Config.load_config()
 
-    def init(self, capsys):
+    def init(self, capsys: CaptureFixture) -> None:
         self._stdout = None
         self._stderr = None
         self._capsys = capsys
         self.types = None
         self.version = None
         self.resources = {}
-        self._root_scope = {}
+        self._root_scope = None
```

Comment:
This is a breaking change in the senses that `get_root_scope()` can return None now. I am not sure whether that's an issue.

Reply:
The rootscope is a NameSpace object, so it doesn't really matter I think